### ISSUE TITLE
Allstate Claims Severity Competition 2nd Place Winner's Interview: Alexey Noskov

### ISSUE THREAD

**Issue by lucaswadedavis** [lucaswadedavis/reading-list#21]
Allstate Claims Severity Competition, 2nd Place Winner's Interview: Alexey Noskov<br><br>
http://ift.tt/2mwaRSN<br><br><img height="280" alt="Allstate Claims Severity recruiting Kaggle competition 2nd place" width="1000" class="wp-post-image" src="http://ift.tt/2l51LiZ"><p>The <a target="_blank" href="http://ift.tt/2dUspoA">Allstate Claims Severity recruiting competition</a> ran on Kaggle from October to December 2016. As Kaggle's most popular recruiting competitions to-date, it attracted over 3,000 entrants who competed to predict the loss value associated with Allstate insurance claims.</p>
<p>In this interview, <a target="_blank" href="http://ift.tt/2mwnfCu">Alexey Noskov</a> walks us through how he came in second place by creating features based on distance from cluster centroids and applying newfound intuitions for (hyper)-parameter tuning. Along the way, he provides details on his favorite tips and tricks including lots of feature engineering and implementing a custom objective function for XGBoost.</p>
<p><img height="25" width="1000" class="wp-image-6230" src="http://ift.tt/2e9JixE"></p>
<h2>Background</h2>
<p>I have MSc in computer science and work as a software engineer at <a href="http://ift.tt/1FKpCHV">Evil Martians</a>.</p>
<div class="wp-caption">
<a href="http://ift.tt/2mwnfCu"><img height="477" alt="Alexey on Kaggle." width="1000" class="wp-image-6509" src="http://ift.tt/2mwoqBz"></a>
<p class="wp-caption-text"><a target="_blank" href="http://ift.tt/2mwnfCu">Alexey</a> on Kaggle.</p>
</div>
<p>I became interested in data science about 4 years ago - first I watched <a href="http://ift.tt/1aQjwZG">Andrew Ng&rsquo;s famous course</a>, then some others, but I lacked experience with real problems and struggled to get some. But things changed when around beginning of 2015 I got to know Kaggle, which seem to be the missing piece, as it allowed me to get experience in complex problems and learn from the others, improving my data science and machine learning skills.</p>
<p>So, for two years already I&rsquo;ve participated in Kaggle competitions as much as I can, and it&rsquo;s one of the most fun and productive pursuits I&rsquo;ve had.</p>
<p>I noticed this competition during the end of <a href="http://ift.tt/2bjYnKJ">Bosch Production Line Performance</a>, and I became interested in it because of the moderate data size and mangled data, so I can focus on general methods of building and improving models. So I entered it as soon as I got some time.</p>
<h2>Data preprocessing and feature engineering</h2>
<p>First, I needed to fix skew in target variable. Initially I applied log-transform, and it worked good enough, but some time after I switched to other transformations like <code>log(loss + 200)</code> or <code>loss ^ 0.25</code>, which worked somewhat better.</p>
<div class="wp-caption">
<img height="453" alt="Target variable and its transformations" width="1006" class="wp-image-6506" src="http://ift.tt/2mwgaBP"><p class="wp-caption-text">Target variable and its transformations.</p>
</div>
<p>For features - first of all, I needed to encode categorical variables. For this I used basic one-hot encoding for some models, but also so-called lexical encoding, when value of encoded category is produced from its name (A becomes 0, B - 1, Z - 26, AA - 27, and so on).</p>
<p>I tried to find some meaningful features, but had no success at it. Also there were <a href="http://ift.tt/2dUspoA">some kernels</a> which provided insights into the nature of some variables, and tried to de-mangle them but I couldn&rsquo;t get any improvement from it. So, I switched to using general automated methods.</p>
<p>The first of such methods was, of course, SVD, which I&rsquo;ve applied to numerical variables and one-hot encoded categorical features. It helped to improve some high-variance models, like FM and NN.</p>
<p>Second, and more complex, was clustering the data and creating a new set of features based on the distance to cluster centers (i.e., applying <a href="http://ift.tt/2mwm45K">RBF</a> to them) - it helped to create a bunch of unsupervised non-linear features, which helped to improve most of my models.</p>
<p>And third, the last trick I used was forming categorical interaction features, applying lexical encoding to them. These combinations may be easily extracted from XGBoost models by just trying the most important categorical features, or better, analysing the model dump with the excellent <a href="http://ift.tt/1S7PFQs">Xgbfi</a> tool.</p>
<h2>First-level models</h2>
<p>Based on these features, I built a lot of different models which I evaluated using the usual k-fold cross-validation.</p>
<p>First of all, there was linear regression, which gave me about 1237.43406 CV / 1223.28163 LB score, which is not very much of course, but provides some baseline. But after adding cluster features to it, it became 1202.70592 CV / 1189.64998 LB, which is much better for such a simple model.</p>
<p>Then, I tried scikit-learn <a href="http://ift.tt/1NsiSik">RandomForestRegressor</a> and <a href="http://ift.tt/2l4ZWm0">ExtraTreesRegressor</a> models, of which random forest was best, giving 1199.82233 CV / 1176.44433 LB after some tuning, and improved to 1186.23675 CV / 1166.85340 LB after adding categorical feature combinations. One problem with this model was that however scikit-learn supports MAE loss, it&rsquo;s very slow and impossible to use, so I needed to use basic MSE, which has some bias in this competition.</p>
<p>The best model of scikit-learn which helped me was <a href="http://ift.tt/1sBqR6p">GradientBoostingRegressor</a>, which was able to directly optimize MAE loss and gave me 1151.11060 CV / 1126.30971 LB</p>
<p>I also tried <a href="http://ift.tt/2l4Vwf1">LibFM</a> model, which gave me 1196.11333 CV / 1155.68632 LB in a basic version and 1177.69251 CV / 1150.37290 LB after adding cluster features to it.</p>
<p>But the main workhorses of this competitions were, of course, <a href="http://ift.tt/2cY7rYH">XGBoost</a> and neural net models:</p>
<p>In the beginning, my XGBoost models provided something about 1133.00048 CV / 1112.86570 LB, but then I&rsquo;ve applied some tricks which improved it to 1122.64977 CV / 1105.43686 LB:</p>
<ul>
<li>Averaging multiple runs of XGBoost with different seeds - it helps to reduce model variance;</li>
<li>Adding categorical combination features;</li>
<li>Modifying objective function to be closer to MAE;</li>
<li>Tuning model parameters - I didn&rsquo;t have much experience with it before, so <a href="http://ift.tt/2l4S1VL">this thread in Kaggle forums</a> helped me a lot.</li>
</ul><div class="wp-caption">
<img height="358" alt="Custom objective function for XGBoost." width="636" class="wp-image-6505" src="http://ift.tt/2mwmyJm"><p class="wp-caption-text">Custom objective function for XGBoost.</p>
</div>
<p>The other model that provided great results was neural net, implemented using the <a href="https://keras.io/">Keras</a> library. I used basic multi-layer perceptron with 3 hidden layers which gave me somewhat about 1134.92794 CV / 1116.44915 LB in initial versions and improved to 1130.29286 CV / 1110.69527 LB after tuning and applying some tricks:</p>
<ul>
<li>Averaging multiple runs, again;</li>
<li>Applying exponential moving average to weights of single network, using <a href="http://ift.tt/2l4URtM">this implementation</a>;</li>
<li>Adding SVD and cluster features;</li>
<li>Adding batch normalization and dropout;</li>
</ul><h2>Model tuning</h2>
<p>In this competition, model hyperparameter tuning was very important, so I&rsquo;ve contributed a lot of time in it. There are three main approaches here:</p>
<ul>
<li>Manual tuning, which works good when you have some intuition about parameter behaviour and may estimate model performance before its training completes by per-epoch validation scores;</li>
<li>Uninformed parameter search - using <a href="http://ift.tt/2fPwlFG">GridSearchCV</a> or <a href="http://ift.tt/2l5188Z">RandomizedSearch</a> from sklearn package, or similar - most simple of all;</li>
<li>Informed search using <a href="http://ift.tt/1qa5FT3">HyperOpt</a> or <a href="http://ift.tt/1PGCqj2">BayesOptimization</a> or similar package - it tries to fit some model to scores of different parameter sets and selects the most promising point for each next try - so it usually finds the optimum a lot faster than uninformed search.</li>
</ul><p>I used manual tuning for XGBoost and NN models which provide per-epoch validation scores and bayes optimization package for the others.</p>
<h2>Second level</h2>
<p>After getting a lot of models, I combined them in the second level, training new models on out-of-fold predictions:</p>
<ul>
<li>Linear regression, which gave me 1118.45564 CV / 1113.08059 LB score</li>
<li>XGBoost - 1118.16984 CV / 1100.50998 LB</li>
<li>Neural net - 1116.40752 CV / 1098.91721 LB (it was enough to get top-16 in public, and top-8 in private)</li>
<li>Gradient boosting - 1117.41247 CV / 1099.60251 LB</li>
</ul><p>I haven&rsquo;t had much experience with stacking before and so I was really impressed by these results, but wanted to get even more.</p>
<p>So, the first thing I did was correct the bias of some stacked models - linear regression and XGBoost optimized some objective which was not equal to the objective of the competition, which resulted in overestimating low values and underestimating high ones. This bias is really small, but the competition was very close, so every digit counted.</p>
<p>This bias can be seen on the next figure where logs of XGBoost predictions are plotted against target logs with and a median regression line. If it was unbiased, the median regression should be the same as diagonal, but it's not (offset is most visible where red arrows are).</p>
<div class="wp-caption">
<img height="362" alt="XGBoost bias." width="758" class="wp-image-6508" src="http://ift.tt/2mwklgI"><p class="wp-caption-text">XGBoost bias.</p>
</div>
<p>I took XGBoost predictions to some small power <code>p</code> (around 1.03), normalized to preserve its median, and it improved my score to 1117.35084 CV / 1099.63060 LB.</p>
<p>Not bad, but maybe I can get even more by combining predictions of these models?</p>
<h2>Third level</h2>
<p>So, I built a third level. As each new stacking level becomes more and more unstable, I needed something really simple here, which may optimize the competition&rsquo;s metric directly. So, I chose to use median regression from the <a href="http://ift.tt/1pXqFYs">statsmodels</a> package.</p>
<p>The main problem of this approach was lack of regularization, so it wasn&rsquo;t very stable and had a lot of noise. To fight it I applied some tricks:</p>
<ul>
<li>Training model on many subsamples and averaging predictions;</li>
<li>Reducing input dimensionality - grouping similar models of previous layers and using group averages as features;</li>
<li>Averaging best 10 submissions for a final one.</li>
</ul><p>It allowed me to get 1098.07061 in public LB, and 1110.01364 in private, which corresponds to second place.</p>
<div class="wp-caption">
<img height="333" alt="Final pipeline." width="724" class="wp-image-6507" src="http://ift.tt/2mwngX0"><p class="wp-caption-text">Final pipeline.</p>
</div>
<h2>Lessons learned</h2>
<p>So, this competitions helped me a lot, mainly in two main areas in which I lacked experience in before - model hyperparameter tuning (especially for XGBoost, where I got good hyperparameter intuition) and stacking, which I underestimated before.</p>
<p>Also, I tried a lot of different models and got intuition about how they work with different target, feature transformations, and so on.</p>
<p><img height="25" width="1000" class="wp-image-6230" src="http://ift.tt/2e9JixE"></p>
<h2>Bio</h2>
<p><a href="http://ift.tt/2mwnfCu">Alexey Noskov</a> is a Ruby and Scala developer at Evil Martians.</p>
<h3>More from Alexey</h3>
<a href="http://ift.tt/2l50t7n"><img height="80" width="150" class="wp-image-6491" src="http://ift.tt/2l8qija"></a>
<p>Alexey <a href="http://ift.tt/2l50t7n">shares more details</a> on his winning approach on the competition&rsquo;s forums including his winning competition code <a href="http://ift.tt/2l594XN">on GitHub</a>.</p>
<br><br><br><br>
Data Science<br><br>
via no free hunch http://blog.kaggle.com<br><br>
February 27, 2017 at 04:19AM
